src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -351,109 +360,151 @@ internal override async ValueTask<int> ReadAsync(Memory<byte> destination, Cance
+                abortError = _state.ReadErrorCode;
+
+                if (readState != ReadState.PendingRead && cancellationToken.IsCancellationRequested)
                 {
-                    throw ThrowHelper.GetStreamAbortedException(_state.ReadErrorCode);
+                    readState = ReadState.Aborted;
+                    _state.ReadState = ReadState.Aborted;
+                    canceledSynchronously = true;
                 }
-                else if (_state.ReadState == ReadState.ConnectionClosed)
+                else if (readState == ReadState.None)
                 {
-                    throw GetConnectionAbortedException(_state);
-                }
-            }
+                    Debug.Assert(_state.RootedReceiveStream is null);
 
-            using CancellationTokenRegistration registration = cancellationToken.UnsafeRegister(static (s, token) =>
-            {
-                var state = (State)s!;
-                bool shouldComplete = false;
-                lock (state)
-                {
-                    if (state.ReadState == ReadState.None)
+                    _state.ReceiveUserBuffer = destination;
+                    _state.RootedReceiveStream = this;
+                    _state.ReadState = ReadState.PendingRead;
+
+                    if (cancellationToken.CanBeCanceled)
                     {
-                        shouldComplete = true;
+                        _state.ReceiveCancellationRegistration = cancellationToken.UnsafeRegister(static (obj, token) =>
+                        {
+                            var state = (State)obj!;
+                            bool completePendingRead;
+
+                            lock (state)
+                            {
+                                completePendingRead = state.ReadState == ReadState.PendingRead;
+                                state.RootedReceiveStream = null;
+                                state.ReceiveUserBuffer = null;
+                                state.ReadState = ReadState.Aborted;
+                            }
+
+                            if (completePendingRead)
+                            {
+                                state.ReceiveResettableCompletionSource.CompleteException(ExceptionDispatchInfo.SetCurrentStackTrace(new OperationCanceledException(token)));
+                            }
+                        }, _state);
+                    }
+                    else
+                    {
+                        _state.ReceiveCancellationRegistration = default;
                     }
-                    state.ReadState = ReadState.Aborted;
-                }
 
-                if (shouldComplete)
-                {
-                    state.ReceiveResettableCompletionSource.CompleteException(
-                        ExceptionDispatchInfo.SetCurrentStackTrace(new OperationCanceledException("Read was canceled", token)));
+                    return _state.ReceiveResettableCompletionSource.GetValueTask();
```

Comment:
The combination of `if-else if-else if` with `return` at the end of some of the branches is super confusing to me. For instance, the first `if (readState ...)` will continue after the whole block, but the two `else if-else if` both end with `return`, ending the flow there. Could we reshuffle this? Maybe put the two else-if which return as first and make them just ifs.

What is your opinion on the code organization of this? Do you find it easily readable/comprehensible? Maybe I'm just not familiar enough with this style.

Reply:
As we just talked offline, feel free to keep this as is, even for a follow up. My "feelings" about readability are not sound enough reason for so much work.